src/test/ui/closures/issue-101696.rs

```diff
@@ -0,0 +1,36 @@
+// check-pass
+
+use std::marker::PhantomData;
+
+#[derive(Default)]
+struct MyType<'a> {
+    field: usize,
+    _phantom: PhantomData<&'a ()>,
+}
+
+#[derive(Default)]
+struct MyTypeVariant<'a> {
+    field: usize,
+    _phantom: PhantomData<&'a ()>,
+}
+
+trait AsVariantTrait {
+    type Type;
+}
+
+impl<'a> AsVariantTrait for MyType<'a> {
+    type Type = MyTypeVariant<'a>;
+}
+
+type Variant<G> = <G as AsVariantTrait>::Type;
+
+fn foo<T: Default, F: FnOnce(T)>(f: F) {
+    let input = T::default();
+    f(input);
+}
+
+fn main() {
+    foo(|a: <MyType as AsVariantTrait>::Type| {
```

Comment:
Is it worth adding another test for the notation that tries to use the `type` alias `Variant` instead?

```rust
fn main() {
    foo(|a: Variant<MyType>| { 
        a.field;
    });
}
```

Reply:
Oh, I can just change this test to use that. I only changed this to the associated type syntax to make sure it wasn't a bug with type aliases.